### PR TITLE
feat(proxy): Make TLS configurable for client-to-proxy and proxy-to-server

### PR DESCRIPTION
**Pull request by duesee**
A first attempt of providing configurable TLS support for "both sides". This required fleshing-out our `Config` format and wrapping our connections in a `TlsAcceptor` (client-to-proxy) and `TlsConnector` (proxy-to-server).

* Bypassing of certificate validation is not implemented, and I would argue against implementing it
* STARTTLS is not implemented, and I would argue against implementing it
* PKCS #&#8203;12 support is not implemented yet, but we probably want it at some point
* Encrypted keys (passphrases) are not implemented yet, but we probably want it at some point

~~I made a draft PR to see if CI is happy. But I'd still like to change a few things. Especially, I'd like to evaluate https://crates.io/crates/pem-rfc7468 Thus, not requesting a review yet.~~

### CODE REVIEW

**Review by jakoschiko — COMMENTED** [review]
I'm not suitable for reviewing the certificate stuff, but it's looks nice and clean.

File: proxy/configs/tls_tls.toml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+[services.bind]
+encryption = "Tls"
+host = "127.0.0.1"
+port = 1993
+
+[services.bind.identity]
+type = "CertificateChainAndLeafKey"
+certificate_chain_path = "localhost.pem"
+leaf_key_path = "localhost-key.pem"
```
[the remark this answers is not in the document]
Comment (duesee) [reply proxy/configs/tls_tls.toml:12]: Generally, I highly recommend using https://github.com/FiloSottile/mkcert. But some additional things need to be said (and maybe done). I will open an issue to add a bit of ducumentation.

File: proxy/src/proxy.rs [hunk 2]
```diff
@@ -55,11 +59,44 @@ impl Proxy<BoundState> {
...
+                        } => {
+                            let certificate_chain =
+                                util::load_certificate_chain_pem(certificate_chain_path)?;
+                            let leaf_key = util::load_leaf_key_pem(leaf_key_path)?;
+
+                            (certificate_chain, leaf_key)
+                        }
+                    };
+
+                    // XXX: Support OCSP stapling?
```
[the remark this answers is not in the document]
Comment (duesee) [reply proxy/src/proxy.rs:78]: Sorry, this was confusing. In fact, I don't know if we want to care about this at this point. I prefer to delete this note for now.

File: proxy/src/util.rs [hunk 3]
```diff
@@ -86,3 +90,65 @@ fn filter_capabilities(capabilities: NonEmptyVec<Capability>) -> NonEmptyVec<Cap
...
+
+    rustls_pemfile::certs(&mut reader)
+        .map(|res| {
+            res.map(|der| Certificate(der.to_vec()))
+                .map_err(|source| IdentityError::Io {
+                    source,
+                    path: display_path.clone(),
+                })
+        })
+        .collect()
```
[the remark this answers is not in the document]
Comment (duesee) [reply proxy/src/util.rs:126]: Haha, learned this trick from you :D I think it was also in the Crust of Rust episode on std::collections?

File: proxy/README.md [hunk 4]
```diff
@@ -53,16 +53,20 @@ RUST_LOG=proxy=trace cargo run
 
 ## Config
 
-The config consists of multiple "services", and each service looks roughly like this:
+Have a look at the `configs` folder for different scenarios.
```
[the remark this answers is not in the document]
Comment (duesee) [reply proxy/README.md:56]: > the user might think that the proxy is reading and merging all configs from the configs folder, which is not true.

True and unfortunate. I'll change it :+1:

**Comment by duesee** [comment]
Thanks for the review! I removed the `configs` folder and included all scenarios in `config.toml`. I commented-out a few because they won't work anyway due to missing certificates.

**Review by jakoschiko — APPROVED** [review]

**Pull request merged**